.gitlab-ci.yml

```diff
@@ -0,0 +1,124 @@
+image: golang:1.11-alpine3.9
+
+cache:
+  paths:
+    - /go/src/github.com
+    - /go/src/golang.org
+    - /go/src/google.golang.org
+    - /go/src/gopkg.in
+
+variables:
+   DOCKER_HOST: tcp://docker:2375/
+   DOCKER_DRIVER: overlay2
+   DOCKER_REPOSITORY: mendersoftware/mender-artifact
+
+services:
+  - docker:dind
+
+stages:
+  - test
+  - build
+  - publish
+  - integration
+
+before_script:
+  - mkdir -p /go/src/github.com/mendersoftware /go/src/_/builds
+  - cp -r $CI_PROJECT_DIR /go/src/github.com/mendersoftware/mender-artifact
+  - ln -s /go/src/github.com/mendersoftware /go/src/_/builds/mendersoftware
+  - cd /go/src/github.com/mendersoftware/mender-artifact
+  - apk update && apk add git make
+  # Install code coverage tooling
+  - make get-tools
+
+test:static:
+  stage: test
+  script:
+    - apk add bash e2fsprogs e2fsprogs-extra gcc libc6-compat mtools musl-dev parted perl-utils xz-dev
+    - make extracheck
+    - make coverage
+    - mv /go/src/github.com/mendersoftware/mender-artifact/coverage.txt $CI_PROJECT_DIR/coverage.txt
+  artifacts:
+    untracked: true
+    paths:
+      - coverage.txt
+
+build:
+  image: docker
+  before_script: []
+  stage: build
+  script:
+    - docker build -t $DOCKER_REPOSITORY:pr .
+    - docker save $DOCKER_REPOSITORY:pr > image.tar
+  artifacts:
+    expire_in: 1h
+    paths:
+      - image.tar
+  tags:
+    - docker
+
+build:make:
+  image: golang:1.11
+  before_script: []
+  stage: build
+  script:
+    - apt-get update && apt-get install -yyq liblzma-dev
+    - mkdir -p /go/src/github.com/mendersoftware /go/src/_/builds
+    - cp -r $CI_PROJECT_DIR /go/src/github.com/mendersoftware/mender-artifact
+    - ln -s /go/src/github.com/mendersoftware /go/src/_/builds/mendersoftware
+    - cd /go/src/github.com/mendersoftware/mender-artifact
+    - make get-tools
+    - make build
+    - make extracheck
+    - cp /go/src/github.com/mendersoftware/mender-artifact/mender-artifact $CI_PROJECT_DIR
+    - chmod +x $CI_PROJECT_DIR/mender-artifact
+  artifacts:
+    expire_in: 1h
+    paths:
+      - mender-artifact
+  tags:
+    - docker
+
+publish:tests:
+  stage: publish
+  before_script: []  
+  dependencies:
+    - test:static
+  script:
+    - echo "coverage-acceptance"
+    # Integrate with https://codecov.io
+    # - bash <(curl -s https://codecov.io/bash) -F acceptance -f coverage.txt;
+
+publish:build:
+  image: docker
+  before_script: []
+  stage: publish
+  dependencies:
+    - build
+  script:
+    - echo "publishing image to docker hub"
+    # - docker load -i image.tar
+    # - if [ ! -z "$CI_COMMIT_TAG" ]; then export IMAGE_TAG=$CI_COMMIT_TAG; else export IMAGE_TAG=$CI_COMMIT_REF_NAME; fi ;
+    # - docker tag $DOCKER_REPOSITORY:pr $DOCKER_REPOSITORY:$IMAGE_TAG ;
+    # - docker login --username=$DOCKER_HUB_USERNAME --password=$DOCKER_HUB_PASSWORD ;
+    # - docker push $DOCKER_REPOSITORY:$IMAGE_TAG ;
```

Comment:
Very nice!

Reply:
But we'll need to see if we do this from here or from the Mender QA pipeline

Reply:
yep, we could also only push on tags, as that will be directly connected to the QA steps, but in the QA pipeline we have all the images in one docker login (one repo with login information to care)

Reply:
I think we should not push any images from here, but from Jenkins, since that is where our releases are made. Better to have things in one place when we release.

Reply:
Agree, we should use the QA work to align where we publish what - some repos publish from travis and some rely on Jenkins, some on tags, some on master, ...
If we ever get rid of Jenkins we could move all the build-artifact-publishing to the QA pipeline 😄

Reply:
Yes, perhaps I should say "mender-qa" pipeline, not Jenkins pipeline. The point is that it's the big one that combines the repos.